### PR TITLE
MGMT-9379: Service changes to allow day 2 workers on single node

### PR DESCRIPTION
**Pull request by paul-maidment**
1: When A day2 worker node is trying to join a single node cluster, the
   node was being incorrectly assigned the wrong role (master instead of
worker) this code fixes that problem.

2: Ensure that the High Availability Mode install command parameter is
   not mandatory.

It used to be the case that a day 2 worker could only be added to an HA
cluster. We are making this change to allow the addition of a day 2
worker to a non HA cluster.

The condition [1] was blocking this.

That validation ensures that Day 1 SNO clusters will only accept Master
or Bootsrap roles for the host being installed.

So an ideal solution was to pass an empty string for "HA mode" whenever
the role of the host is "Worker" as when adding a worker "High
Availability" mode is not relevant. (Especially when sometimes it is not
possible to know the true High Availability Mode of the cluster. For
example during a SaaS Day 2 installation.)

We have only changed the API to allow us to not pass the HA mode
parameter. Presently the assisted-service will still pass a non nil
value for High Availability Mode for workers. We need to bump the 
service dependency of the agent to use this new api change before the
service can omit the HA mode for workers.

[1] https://github.com/openshift/assisted-installer/blob/d67d2843d0e757ef5643e6c873b49e61ef83a0ea/src/config/config.go#L93-L99

## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [x] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by omertuc** [comment]
image

Please explain the problem that's being solved by this PR, and in short how it's being solved

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -4288,13 +4288,17 @@ func (b *bareMetalInventory) V2RegisterHost(ctx context.Context, params installe
...
 			}
 		}
-		if common.IsSingleNodeCluster(cluster) {
+
+		if common.IsDay2Cluster(cluster) {
+			host.Kind = swag.String(models.HostKindAddToExistingClusterHost)
+			host.Role = models.HostRoleWorker
+			host.MachineConfigPoolName = string(models.HostRoleWorker)
+		} else if common.IsSingleNodeCluster(cluster) {
+			// The question of whether or not the cluster is single node or not only matters for a Day 1 installation.
```
Comment (omertuc) [inline internal/bminventory/inventory.go:4297]:
```suggestion
			// The question of whether or not the host's cluster is single node or not only matters for a Day 1 installation.
```

File: internal/common/common.go [hunk 2]
```diff
@@ -115,6 +115,11 @@ func IsSingleNodeCluster(cluster *Cluster) bool {
 	return swag.StringValue(cluster.HighAvailabilityMode) == models.ClusterHighAvailabilityModeNone
 }
 
+// IsSingleNodeCluster if this cluster is Day 2 or not
```
Comment (omertuc) [inline internal/common/common.go:118]: doc doesn't match function name

File: internal/controller/controllers/agent_controller.go [hunk 3]
```diff
@@ -77,6 +77,37 @@ type AgentReconciler struct {
 	ApproveCsrsRequeueDuration time.Duration
 }
 
+func (r *AgentReconciler) getClusterDeployment(agent *aiv1beta1.Agent, context *context.Context) (*hivev1.ClusterDeployment, error) {
+	kubeKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
```
Comment (omertuc) [inline internal/controller/controllers/agent_controller.go:82]: You shouldn't access `agent.Spec.ClusterDeploymentName` before you know it's not `nil`

File: internal/controller/controllers/agent_controller.go [hunk 4]
```diff
@@ -77,6 +77,37 @@ type AgentReconciler struct {
...
+			return clusterDeployment, nil
+		}
+		return nil, err
+	}
+	return nil, nil
+}
+
+func (r *AgentReconciler) getCluster(agent *aiv1beta1.Agent) (*common.Cluster, error){
+	kubeKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
```
Comment (omertuc) [inline internal/controller/controllers/agent_controller.go:98]: You shouldn't access agent.Spec.ClusterDeploymentName before you know it's not nil

File: internal/controller/controllers/agent_controller.go [hunk 5]
```diff
@@ -77,6 +77,37 @@ type AgentReconciler struct {
...
+}
+
+func (r *AgentReconciler) getCluster(agent *aiv1beta1.Agent) (*common.Cluster, error){
+	if agent.Spec.ClusterDeploymentName != nil {
+		kubeKey := types.NamespacedName{
+			Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+			Name:      agent.Spec.ClusterDeploymentName.Name,
+		}
+		cluster, err :=  r.Installer.GetClusterByKubeKey(kubeKey)
+		if (err == nil) {
```
Comment (omertuc) [inline internal/controller/controllers/agent_controller.go:103]: `Go` doesn't require parenthesis around if statement conditions

File: internal/controller/controllers/agent_controller.go [hunk 6]
```diff
@@ -157,29 +188,23 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
...
-		kubeKey := types.NamespacedName{
-			Namespace: agent.Spec.ClusterDeploymentName.Namespace,
-			Name:      agent.Spec.ClusterDeploymentName.Name,
-		}
-		clusterDeployment := &hivev1.ClusterDeployment{}
-
-		// Retrieve clusterDeployment
-		if err = r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		_, err := r.getClusterDeployment(agent, &ctx)
+		if (err != nil) {
```
Comment (omertuc) [inline internal/controller/controllers/agent_controller.go:192]: Go doesn't require parenthesis around if statement conditions

File: internal/controller/controllers/agent_controller.go [hunk 7]
```diff
@@ -440,13 +481,26 @@ func (r *AgentReconciler) updateStatus(ctx context.Context, log logrus.FieldLogg
 			agent.Status.ValidationsInfo = newValidationsInfo
 		}
 
-		if h.Progress != nil && h.Progress.CurrentStage != "" {
-			if isNoneDay2Rebooting, err = r.isDay2NonePlatformHostRebooting(ctx, agent, h); err != nil {
-				log.WithError(err).Errorf("Failed to find if agent %s/%s belongs to none platform cluster and is rebooting", agent.Namespace, agent.Name)
+		// Determine whether or not CSRs should be automatically signed.
+		if (cluster != nil) {
```
Comment (omertuc) [inline internal/controller/controllers/agent_controller.go:485]: Go doesn't require parenthesis around if statement conditions

File: internal/controller/controllers/agent_controller.go [hunk 8]
```diff
@@ -440,13 +481,26 @@ func (r *AgentReconciler) updateStatus(ctx context.Context, log logrus.FieldLogg
...
+			isSNOCluster := *cluster.HighAvailabilityMode == string(HighAvailabilityModeNone)
+			isDay2Host := r.isDay2Host(ctx, agent, h)
+			isNonePlatformCluster, err := r.isNonePlatformCluster(ctx, agent, h)
+			if (err != nil) {
+				log.WithError(err).Errorf("Failed to find if agent %s/%s belongs to none platform cluster", agent.Namespace, agent.Name)
+				return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, nil
+			}
+			isRebooting, err := r.isRebootingHost(ctx, agent, h)
+			if (err != nil) {
```
Comment (omertuc) [inline internal/controller/controllers/agent_controller.go:494]: Go doesn't require parenthesis around if statement conditions

File: internal/controller/controllers/agent_controller.go [hunk 9]
```diff
@@ -393,30 +428,36 @@ func (r *AgentReconciler) deregisterHostIfNeeded(ctx context.Context, log logrus
...
-		h.Progress.CurrentStage == models.HostStageDone {
-		if agent.Status.Progress.CurrentStage == models.HostStageDone {
-			return false, nil
-		} else {
-			isNone, err := isAgentInNonePlatformCluster(ctx, r.Client, agent)
-			if err != nil {
-				return false, err
-			}
-			return isNone, nil
+func (r *AgentReconciler) isRebootingHost(ctx context.Context, agent *aiv1beta1.Agent, h *models.Host) (bool, error) {
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/controller/controllers/agent_controller.go:431]: For day2 host, it moves to Done when it starts rebooting. Here https://github.com/openshift/assisted-service/blob/b31fbf34d02bf54e278276af53cea6a3e484baa6/internal/host/host.go#L571
Better call it isDay2HostRebooting because it is irrelevant to day1

[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/controller/controllers/agent_controller.go:431]: And please return the check for day2 
**if swag.StringValue(h.Status) == models.HostStatusAddedToExistingCluster**

File: internal/controller/controllers/agent_controller.go [hunk 10]
```diff
@@ -440,13 +481,26 @@ func (r *AgentReconciler) updateStatus(ctx context.Context, log logrus.FieldLogg
...
+		// Determine whether or not CSRs should be automatically signed.
+		if (cluster != nil) {
+			isSNOCluster := *cluster.HighAvailabilityMode == string(HighAvailabilityModeNone)
+			isDay2Host := r.isDay2Host(ctx, agent, h)
+			isNonePlatformCluster, err := r.isNonePlatformCluster(ctx, agent, h)
+			if (err != nil) {
+				log.WithError(err).Errorf("Failed to find if agent %s/%s belongs to none platform cluster", agent.Namespace, agent.Name)
+				return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, nil
+			}
+			isRebooting, err := r.isRebootingHost(ctx, agent, h)
```
Comment (ori-amizur) [inline internal/controller/controllers/agent_controller.go:493]: The logic that was here before is more correct, because the rebooting logic is only relevant to day2

File: internal/host/hostcommands/install_cmd.go [hunk 11]
```diff
@@ -114,6 +114,11 @@ func (i *installCmd) getFullInstallerCommand(cluster *common.Cluster, host *mode
 		BootDevice:           swag.String(bootdevice),
 	}
 
+	// The HA mode is not relevant for workers.
+	if string(role) == models.HostUpdateParamsHostRoleWorker {
```
Comment (ori-amizur) [inline internal/host/hostcommands/install_cmd.go:118]: Is this change required ? how did it work before ?

Comment (omertuc) [reply internal/host/hostcommands/install_cmd.go:118]: The assisted-installer's CLI param validation checks that if the HA mode is set to "None" then the role has to be master/bootstrap, which (currently) makes sense in day-1 - HA mode means SNO, SNO means there's just one node, and that node has to be master/bootstrap.

But for workers, in general, it doesn't matter if they're day-1 or day-2 - the HA mode doesn't make any difference. Workers don't care what kind of cluster they're going to join. 

How it worked so far -
- In ZTP, we have the HA mode.
- In REST, we don't have the HA mode (it's empty).

You can see on line 100 when `cluster.HighAvailabilityMode` is `nil` we just default to assuming it's `models.ClusterHighAvailabilityModeFull`, which caused the CLI param validation I mentioned in the beginning to be skipped.

But in ZTP, and in SNO, that validation started failing, because we're trying to get a worker to join a non-HA cluster.

@paul-maidment made a similar change on the installer and agent to accept a `nil` HA mode - this is useful for workers where we don't care about the HA. These changes would have to be merged before this PR does

**Comment by omertuc** [comment]
Don't forget to modify the title of this PR to reflect the new scope of the changes
